README.rst
============================
MATRIXIO Python MALOS Driver
============================
A simple `Python`_ coroutine based driver for communicating with `MATRIX-MALOS services`_.

License
=======

This application follows the GNU General Public License, as described in the ``LICENSE`` file.

Installing
==========

The package is available on PyPI, so you can easily install via pip:

.. code-block:: console

    $ pip install matrix-io-malos


Running the CLI client
======================

The library includes a simple command line client to start reading data from 
your MALOS service right away. 

.. code-block:: console

    # Get the malosclient help screen
    $ malosclient --help

    # Get IMU data to STDOUT from a locally running MALOS service
    $ malosclient IMU

    # Get HUMIDITY data to STDOUT from a remotely running MALOS service
    $ malosclient -h 192.168.0.100 HUMIDITY

    # Get FACE detection data using a serialized driver config file
    $ malosclient --driver-config-file ~/driver_config.proto VISION


Using the MalosDriver
=====================

To use the MALOS driver works as an async generator so in your code 
you can do the following:

.. code-block:: python

    import asyncio
    import sys

    from matrix_io.malos.driver import IMU_PORT, UV_PORT
    from matrix_io.proto.malos.v1 import driver_pb2
    from matrix_io.proto.malos.v1 import sense_pb2

    from matrix_io.malos.driver import MalosDriver


    async def imu_data(imu_driver):
        async for msg in imu_driver.get_data():
            print(sense_pb2.Imu().FromString(msg))
            await asyncio.sleep(1.0)


    async def uv_data(uv_driver):
        async for msg in uv_driver.get_data():
            print(sense_pb2.UV().FromString(msg))
            await asyncio.sleep(1.0)


    async def status_handler(driver):

        type_mapping = {
            driver_pb2.Status.MESSAGE_TYPE_NOT_DEFINED: "Not Defined",
            driver_pb2.Status.STARTED: "Started",
            driver_pb2.Status.STOPPED: "Stopped",
            driver_pb2.Status.CONFIG_RECEIVED: "Config Received",
            driver_pb2.Status.COMMAND_EXECUTED: "Command Executed",
            driver_pb2.Status.STATUS_CRITICAL: "Critical",
            driver_pb2.Status.STATUS_ERROR: "Error",
            driver_pb2.Status.STATUS_WARNING: "Warning",
            driver_pb2.Status.STATUS_INFO: "Info",
            driver_pb2.Status.STATUS_DEBUG: "Debug"
        }

        async for msg in driver.get_status():
            print(type_mapping[msg.type])

            if msg.uuid:
                print("UUID: {}".format(msg.uuid))
            if msg.message:
                print("MESSAGE: {}".format(msg.message))

            await asyncio.sleep(1.0)


    # Driver configuration
    driver_config = driver_pb2.DriverConfig()

    # Create the drivers
    imu_driver = MalosDriver('localhost', IMU_PORT)
    uv_driver = MalosDriver('localhost', UV_PORT)

    # Create loop and initialize keep-alive
    loop = asyncio.get_event_loop()
    loop.run_until_complete(imu_driver.configure(driver_config))
    loop.run_until_complete(uv_driver.configure(driver_config))
    loop.create_task(imu_driver.start_keep_alive())
    loop.create_task(uv_driver.start_keep_alive())

    # Initialize data and error handlers
    loop.create_task(imu_data(imu_driver))
    loop.create_task(uv_data(uv_driver))
    loop.create_task(status_handler(imu_driver))
    loop.create_task(status_handler(uv_driver))

    try:
        loop.run_forever()
    except KeyboardInterrupt:
        print('Shutting down. Bye, bye !', file=sys.stderr)
    finally:
        loop.stop()
        asyncio.gather(*asyncio.Task.all_tasks()).cancel()

        loop.run_until_complete(loop.shutdown_asyncgens())
        loop.close()

Who can answer questions about this library?
============================================

- Heitor Silva <heitor.silva@admobilize.com>
- Maciej Ruckgaber <maciek.ruckgaber@admobilize.com>

More Documentation
==================

.. toctree::
    :titlesonly:

    CHANGELOG

.. _0MQ: http://zeromq.org/
.. _Python: https://www.python.org/
.. _virtualenv: https://virtualenv.pypa.io/en/stable/
.. _matrixio-protos-0.0.25: https://pypi.org/project/matrix-io-proto
.. _pypi: https://pypi.org/
.. _MATRIX-MALOS services: https://matrix-io.github.io/matrix-documentation/matrix-core/getting-started/understanding-core/



CHANGELOG.md
# Changelog
MATRIXIO Python MALOS Driver project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [0.4.1] - 2019-04-24
### Changed
- Updated pyzmq dependency from version 17.1.2 to 18.0.1.

## [0.4.0] - 2019-04-23
### Added
- `get_frame` asynchronous generator to handle
  video streaming from Malos.

## [0.3.2] - 2018-11-14
### Changed
- Set matrix-io-proto to be at least version 0.0.29

## [0.3.1] - 2018-11-08
### Changed
- `MalosDriver.configure` timeout now waits for the timeout asynchronously

### Added
- New Status message types to example script

## [0.3.0] - 2018-11-07
### Added
- Timeout added to `MalosDriver.configure` function

### Changed
- `MalosDriver.configure` is an async function now
- Update matrix-io-proto from 0.0.26 to 0.0.29

### Removed
- `zmq.LINGER` time reduced from 3000 ms to 0 ms on `MalosDriver` sockets

## [0.2.2] - 2018-10-17
### Changed
- Drone publish configuration.

## [0.2.1] - 2018-10-17
### Added
- Publish package to pypi on tag event.

## [0.2.0] - 2018-10-02

### Added
- `get_status` function that returns a `driver_pb2.Status` object

### Changed
- Update matrix-io-proto from 0.0.26 to 0.0.27
- `error` port became `status` port
- Keep Alive port now raises `MalosKeepAliveTimeout` if Malos doesn't respond with
pong in a certain period defined by the `timeout` parameter
- Keep Alive now raises `asyncio.CancelledError` when cancelled

### Removed
- Print whole `DriverConfig` on debug log when calling `configure`

## [0.1.1] - 2018-08-31

### Added
- Set LINGER option on `MalosDriver` zmq context to avoid hanging 
when closing application 
- Close all sockets after their task is cancelled

### Changed
- Update matrix-io-proto from 0.0.25 to 0.0.26
- Update pyzmq from 17.0.0 to 17.1.2
- `MalosDriver.configure()` is not called on initialization anymore

### Removed
- `MalosDriver.__init__()` `config_proto` parameter

## [0.1.0] - 2018-08-23
### Added
- `MalosDriver` class for encapsulation of driver functions

### Changed
- Change data and error handlers from callback style to asynchronous 
generators

### Removed
- `MalosLoop` class

## [0.0.4] - 2018-05-14
### Added
- MALOS Driver 0MQ interface functions
- MalosLoop class for starting drivers
- `malosagent` command line interface


tests/drivers_test.py
import asyncio
import unittest
from unittest.mock import ANY

import zmq
from unittest import mock
from zmq.asyncio import Context

from matrix_io.malos.drivers import driver_port, keep_alive_port, error_port, run_driver, data_port
from .asyncio_test_helper import _run, AsyncMock, async_print
from matrix_io.proto.malos.v1 import driver_pb2

# Driver configuration
driver_config = driver_pb2.DriverConfig()

class DriverTest (unittest.TestCase):
    
    host = "localhost"
    port = 0

    @mock.patch('zmq.asyncio.Socket.connect')
    def test_driver_port_connect(self, connect):
        driver_port(self.host, self.port, Context(), driver_config)
        assert connect.called
        connect.assert_called_with('tcp://{0}:{1}'.format(self.host, self.port))


    @mock.patch('zmq.asyncio.Socket.send')
    def test_driver_port_send(self, send):
        driver_port(self.host, self.port, Context(), driver_config)
        assert send.called
        send.assert_called_with(driver_config.SerializeToString())

    @mock.patch('zmq.asyncio.Socket.close')
    def test_driver_port_close(self, close):
        driver_port(self.host, self.port, Context(), driver_config)
        assert close.called

    @mock.patch('zmq.asyncio.Socket.connect')
    @mock.patch('zmq.asyncio.Socket.send_string', new=AsyncMock(side_effect=asyncio.CancelledError()))
    def test_keep_alive_port(self, connect):
        from zmq.asyncio import Socket
        try:
            _run(keep_alive_port(self.host, self.port, Context()))
            connect.assert_called_with('tcp://{0}:{1}'.format(self.host, self.port+1))
            assert Socket.send_string.mock.called
        except asyncio.CancelledError:
            self.fail("keep_alive_port shouldn't raise exception")

    @mock.patch('zmq.asyncio.Socket.setsockopt')
    @mock.patch('zmq.asyncio.Socket.connect')
    @mock.patch('zmq.asyncio.Socket.recv_multipart', new=AsyncMock(side_effect=Exception()))
    def test_error_port(self, connect, setsockopt):
        from zmq.asyncio import Socket

        try:
            _run(error_port(self.host, self.port, Context(), async_print))
            self.fail("error_port should have raised exception")
        except Exception:
            connect.assert_called_with('tcp://{0}:{1}'.format(self.host, self.port + 2))
            setsockopt.assert_called_with(zmq.SUBSCRIBE, b'')
            assert Socket.recv_multipart.mock.called

    @mock.patch('zmq.asyncio.Socket.setsockopt')
    @mock.patch('zmq.asyncio.Socket.connect')
    @mock.patch('zmq.asyncio.Socket.recv_multipart', new=AsyncMock(side_effect=Exception()))
    def test_data_port(self, connect, setsockopt):
        from zmq.asyncio import Socket

        try:
            _run(data_port(self.host, self.port, Context(), async_print))
            self.fail("data_port should have raised exception")
        except Exception:
            connect.assert_called_with('tcp://{0}:{1}'.format(self.host, self.port + 3))
            setsockopt.assert_called_with(zmq.SUBSCRIBE, b'')
            assert Socket.recv_multipart.mock.called

    @mock.patch('matrix_io.malos.drivers.driver_port')
    @mock.patch('zmq.asyncio.Socket.recv_multipart', new=AsyncMock(side_effect=Exception()))
    @mock.patch('matrix_io.malos.drivers.keep_alive_port', new=AsyncMock())
    @mock.patch('matrix_io.malos.drivers.error_port', new=AsyncMock())
    @mock.patch('matrix_io.malos.drivers.data_port', new=AsyncMock())
    def test_run_driver(self, driver_port):
        from matrix_io.malos.drivers import keep_alive_port
        from matrix_io.malos.drivers import error_port
        from matrix_io.malos.drivers import data_port
        loop = asyncio.get_event_loop()
        try:
            _run(run_driver(loop, self.host, self.port, driver_config, async_print, async_print))
            self.fail("run_driver should have raised exception")
        except Exception:
            driver_port.assert_called_with(self.host, self.port, ANY, driver_config)
            keep_alive_port.mock.assert_called_once_with(self.host, self.port, ANY)
            error_port.mock.assert_called_once_with(self.host, self.port, ANY, ANY)
            data_port.mock.assert_called_once_with(self.host, self.port, ANY, ANY)



tests/asyncio_test_helper.py
import asyncio
from unittest import mock

async def async_print(msg):
    print(msg)

def _run(coro):
    return asyncio.get_event_loop().run_until_complete(coro)

def AsyncMock(*args, **kwargs):
    m = mock.MagicMock(*args, **kwargs)

    # make the coroutine return MagicMock obj
    async def mock_coro(*args, **kwargs):
        return m(*args, **kwargs)

    # add the mock object to coroutine object to be used later
    mock_coro.mock = m
    return mock_coro

matrix_io/malos/driver.py
"""
MALOS -- Hardware abstraction layer for MATRIX Creator usable via 0MQ

This is an asyncio based implementation of the MALOS drivers. It uses
coroutines to handle communications with 0MQ and provides a higher
level API for client code.

You can find more details about the MALOS protocol here:
https://github.com/matrix-io/matrix-creator-malos

List of base ports
------------------

IMU with port 20013.
Humidity with port 20017.
Everloop with port 20021.
Pressure with port 20025.
UV with port 20029.
MicArray_Alsa with port 20037.
VISION with port 60001


Each port reserves a range of 4 ports that are used for a driver:

* Base port: (driver port) port for sending configuration proto to MALOS driver.
* Keep alive port: port for sending and receiving keep alive messages, so MALOS driver keeps reading data
  from sensors.
* Status port: port yielding MALOS status when they occur.
* Data update port: port yielding sensor data updates.

Video stream is available through port 59999.
"""
import sys

import logging
import asyncio

import zmq
from matrix_io.proto.malos.v1 import driver_pb2
from typing import AsyncIterable
from zmq.asyncio import Context

IMU_PORT = 20013
HUMIDITY_PORT = 20017
EVERLOOP_PORT = 20021
PRESSURE_PORT = 20025
UV_PORT = 20029
MICARRAY_ALSA_PORT = 20037
VISION_PORT = 60001

STREAM_PORT = 59999

class MalosKeepAliveTimeout(Exception):
    pass

class MalosConfigureTimeout(Exception):
    pass

class MalosDriver(object):
    """ Coroutine based MALOS manager """

    def __init__(self, address: str, base_port: int):
        """
        Constructor

        Args:
            address: IP address of the device exposing the MALOS 0MQ sockets
            base_port: MALOS base port to use, see list of base ports above.
            config_proto: a driver.proto containing configuration for the driver
        """
        self.address = address
        self.base_port = base_port

        # 0MQ context
        self.ctx = Context.instance()
        self.logger = logging.getLogger(__name__)
        

        # How long to hold pending messages in memory after
        # closing a socket (in milliseconds).
        # Avoids context hanging indefinitely
        # when destroyed.
        self.ctx.setsockopt(zmq.LINGER, 0)

    async def configure(self, config_proto: driver_pb2.DriverConfig, timeout: int = 5) -> None:
        """
        MALOS configuration

        It sends the provided configuration proto, the driver.proto, to the
        MALOS configuration port to configure the driver.

        Details about the proto structure can be found here:
        https://github.com/matrix-io/protocol-buffers/blob/master/malos/driver.proto

        Args:
            config_proto: a driver.proto containing configuration for the driver
            timeout: timeout in seconds to deliver configuration data to Malos

        Raises:
            MalosConfigureTimeout

        Returns:
            None
        """

        # Set up socket as a push
        sock = self.ctx.socket(zmq.PUSH)

        # Connect to the config port, same base port
        sock.connect('tcp://{0}:{1}'.format(self.address, self.base_port))

        config_string = config_proto.SerializeToString()

        timeout = int(timeout)

        try:
            # Send the configuration
            tracker = await sock.send(config_string, copy=False, track=True)
           
            self.logger.debug(':configure: %r seconds timeout' % timeout)
            # Using asyncio.sleep instead of tracker.wait
            # so that other async tasks can run while it waits
            while timeout > 0:
                await asyncio.sleep(1)
                if tracker.done:
                    break
                timeout -= 1

            if timeout == 0 and not tracker.done:
                self.logger.debug(':configure: timeout reached')
                raise MalosConfigureTimeout()
                
            self.logger.debug(':configure: %r bytes delivered' % sys.getsizeof(config_string))
        except asyncio.CancelledError:
            return	
        finally:
            self.logger.debug(':configure: socket closed')
            sock.close()

    async def start_keep_alive(self, delay: int = 5.0, timeout: int = 5.0) -> None:
        """
        MALOS keep-alive starter

        Connects to the corresponding keep-alive port (base_port +1) given the
        desired base_port. We send keep alive pings and receive pongs so MALOS
        keeps sampling the data from the sensors and we can keep yielding with get_data.

        Args:
            delay: delay between pings in seconds
            timeout: how long to wait for pongs before timeout in seconds

        Raises:
            MalosKeepAliveTimeout

        Yields:
            Doesn't yield anything
        """
        sock = self.ctx.socket(zmq.REQ)
        sock.connect('tcp://{0}:{1}'.format(self.address, self.base_port + 1))

        sock.setsockopt(zmq.RCVTIMEO, int(1000*timeout))

        # If the keep-alive pings stop, MALOS will stop the driver and stop
        # sending updates. Pings are useful to prevent blocked applications
        # from keeping a MALOS driver busy.
        try:
            while True:
                # An empty string is enough to let the driver know we're still
                # listening
                await sock.send_string('')
                self.logger.debug(':keep-alive: ping')

                await sock.recv_string()
                self.logger.debug(':keep-alive: pong')

                await asyncio.sleep(delay)
        except zmq.error.Again:
            raise MalosKeepAliveTimeout()
        except asyncio.CancelledError:
            self.logger.debug(':keep-alive: cancelled')
            # re-raise CancelledError
            raise
        finally:
            self.logger.debug(':keep-alive: socked closed')
            sock.close()


    async def get_status(self) -> AsyncIterable[driver_pb2.Status]:
        """
        MALOS status async generator

        Connects to the corresponding status port (base_port +2) given the desired
        base_port and yields messages.

        Yields:
            Status proto
        """
        sock = self.ctx.socket(zmq.SUB)
        sock.connect('tcp://{0}:{1}'.format(self.address, self.base_port + 2))

        # We send an empty message to let it know we're ready
        sock.setsockopt(zmq.SUBSCRIBE, b'')

        try:
            while True:
                msg = await sock.recv_multipart()
                status = driver_pb2.Status().FromString(msg[0])
                self.logger.debug(':status-port: %s' % status)
                yield status
        except asyncio.CancelledError:
            self.logger.debug(':status-port: cancelled')
            # Exit gracefully when cancelled
            return
        finally:
            self.logger.debug(':status-port: socked closed')
            sock.close()


    async def get_data(self) -> AsyncIterable[bytes]:
        """
        MALOS data async generator

        Connects to the corresponding data port (base_port +3) and yields
        messages received through it.

        Yields:
            Data protobuf sent by MALOS.
        """
        sock = self.ctx.socket(zmq.SUB)
        sock.connect('tcp://{0}:{1}'.format(self.address, self.base_port + 3))

        # We send an empty message to let it know we're ready
        sock.setsockopt(zmq.SUBSCRIBE, b'')

        try:
            while True:
                msg = await sock.recv_multipart()
                self.logger.debug(':data-port: %s' % msg)
                yield msg[0]
        except asyncio.CancelledError:
            self.logger.debug(':data-port: cancelled')
            # Exit gracefully when cancelled
            return
        finally:
            self.logger.debug(':data-port: socked closed')
            sock.close()

    async def get_frame(self) -> AsyncIterable[bytes]:
        """
        MALOS frame async generator

        Connects to the corresponding video stream port (59999) and yields
        raw frames received through it.

        Keepalive must be active and a START_STREAMING action must be sent
        to Malos in order to make it start sending frames.

        Yields:
            Raw frame sent by MALOS.
        """
        sock = self.ctx.socket(zmq.SUB)
        sock.connect('tcp://{0}:{1}'.format(self.address, STREAM_PORT))

        # We send an empty message to let it know we're ready
        sock.setsockopt(zmq.SUBSCRIBE, b'')

        try:
            while True:
                msg = await sock.recv_multipart()
                self.logger.debug(':stream-port: %s' % msg)
                yield msg[0]
        except asyncio.CancelledError:
            self.logger.debug(':stream-port: cancelled')
            # Exit gracefully when cancelled
            return
        finally:
            self.logger.debug(':stream-port: socked closed')
            sock.close()



matrix_io/malos/drivers.py
"""
MALOS -- Hardware abstraction layer for MATRIX Creator usable via 0MQ

This is an asyncio based implementation of the MALOS drivers. It uses
coroutines to handle communications with 0MQ and provides a higher
level API for client code.

You can find more details about the MALOS protocol here:
https://github.com/matrix-io/matrix-creator-malos

List of base ports
------------------

IMU with port 20013.
Humidity with port 20017.
Everloop with port 20021.
Pressure with port 20025.
UV with port 20029.
MicArray_Alsa with port 20037.

Each port reserves a range of 4 ports that are used for a driver:

* Base port: (driver port) port for sending configuration proto to MALOS driver.
* Keep alive port: port for sending keep alive messages, so MALOS driver keeps reading data
  from sensors.
* Error port: port yielding MALOS errors when they occur.
* Data update port: port yielding sensor data updates.

"""

import asyncio
import zmq

from zmq.asyncio import Context

IMU_PORT = 20013
HUMIDITY_PORT = 20017
EVERLOOP_PORT = 20021
PRESSURE_PORT = 20025
UV_PORT = 20029
MICARRAY_ALSA_PORT = 20037


def driver_port(address, base_port, ctx, config_proto):
    """
    MALOS configuration port

    Uses the base port and accepts a configuration proto, the driver.proto
    to configure the driver.

    Details about the proto structure can be found here:
    https://github.com/matrix-io/protocol-buffers/blob/master/malos/driver.proto

    Args:
        address: IP address of the device exposing the MALOS 0MQ sockets
        base_port: desired base port as in List of base ports above (ie. 20013, 20017)
        ctx: a 0MQ context
        config_proto: a driver.proto containing configuration for the driver
    """

    # Set up socket as a push
    sock = ctx.socket(zmq.PUSH)

    # Connect to the config port, same base port
    sock.connect('tcp://{0}:{1}'.format(address, base_port))

    # Send the configuration
    sock.send(config_proto.SerializeToString())
    sock.close()


async def keep_alive_port(address, base_port, ctx, delay=5.0):
    """
    MALOS keep-alive port

    Connects to the corresponding keep-alive port (base_port +1) given the desired
    base_port. The keep alive port will run as long as the data_port is yielding
    messages from 0MQ.

    Args:
        address: IP address of the device exposing the MALOS 0MQ sockets
        base_port: desired base port as in List of base ports above (ie. 20013, 20017)
        ctx: a 0MQ context
        delay: delay between pings
    """

    # Set up socket as a push
    sock = ctx.socket(zmq.PUSH)

    # Connect to the keep alive port to the sensor port from the function args + 1
    sock.connect('tcp://{0}:{1}'.format(address, base_port + 1))

    # If the keep-alive pings stop, MALOS will stop the driver and stop sending
    # updates. Pings are useful to prevent blocked applications from keeping
    # a MALOS driver bussy.
    while True:
        try:
            # Ping with empty string to let the drive know we're still listening
            await sock.send_string('')

            # Delay between next ping
            await asyncio.sleep(delay)
        except asyncio.CancelledError:
            # Exit gracefully
            break


async def error_port(address, base_port, ctx, callback):
    """
    MALOS error port

    Connects to the corresponding error port (base_port +2) given the desired
    base_port and pushes messages to the provided callback.

    Args:
        address: IP address of the device exposing the MALOS 0MQ sockets
        base_port: desired base port as in List of base ports above (ie. 20013, 20017)
        ctx: a 0MQ context
        callback: a callback function to send received messages to
    """

    # Set up socket as a subscription
    sock= ctx.socket(zmq.SUB)

    # Connect to the base sensor port provided in the args + 2 for the error port
    sock.connect('tcp://{0}:{1}'.format(address, base_port + 2))

    # Set socket options to subscribe and send off en empty string to let it know we're ready
    sock.setsockopt(zmq.SUBSCRIBE, b'')

    while True:
        try:
            msg = await sock.recv_multipart()
            await callback(msg)
        except asyncio.CancelledError:
            break


async def data_port(address, base_port, ctx, callback):
    """
    MALOS data update port

    Connects to the corresponding error port (base_port +3) given the desired
    base_port and sends messages to the provided callback.

    Args:
        addr: IP address of the device exposing the MALOS 0MQ sockets
        base_port: desired base port as in List of base ports above (ie. 20013, 20017)
        ctx: a 0MQ context
        callback: a callback function to send received messages to
    """

    # Set up socket as a subscription
    sock = ctx.socket(zmq.SUB)

    # Connect to the base sensor port provided in the args + 3 for the data port
    sock.connect('tcp://{0}:{1}'.format(address, base_port + 3))

    # Set socket options to subscribe and send off en empty string to let it know we're ready
    sock.setsockopt(zmq.SUBSCRIBE, b'')

    while True:
        try:
            msg = await sock.recv_multipart()
            await callback(msg[0])
        except asyncio.CancelledError:
            break


async def run_driver(address, base_port, config_proto, data_callback, error_callback):
    context = Context()

    # Configure the MALOS driver first
    driver_port(address, base_port, context, config_proto)

    # Run the keep-alive channel
    kalive = asyncio.ensure_future(keep_alive_port(address, base_port, context))

    # Run the error channel
    error = asyncio.ensure_future(error_port(address, base_port, context, error_callback))
    try:
        # Run the data update channel
        await data_port(address, base_port, context, data_callback)
    except asyncio.CancelledError:
        kalive.cancel()
        error.cancel()


matrix_io/malos/malosloop.py
import asyncio

from .drivers import run_driver


class MalosLoop:
    def __init__(self):
        self.loop = asyncio.get_event_loop()
        self.drivers = []

    def configure_driver(self, address, base_port, config, data_callback, error_callback=None):
        self.drivers.append(asyncio.ensure_future(
            run_driver(self.loop, address, base_port, config, data_callback, error_callback),
            loop=self.loop))

    def run(self):
        self.loop.run_forever()

    def stop(self):
        # Cancel the all channels for each driver
        for driver in self.drivers:
            driver.cancel()

        # print raised exceptions
        for driver in self.drivers:
            try:
                if driver.done():
                    driver.result()
            except Exception as e:
                print("Unexpected exception: ", e)

        # Cancel driver children
        self.loop.run_until_complete(self.loop.shutdown_asyncgens())
        # Workaround: cancel run_driver. If this line is removed,
        # run_driver pending warning will show up
        self.loop.run_until_complete(self.loop.shutdown_asyncgens())
        self.loop.close()


matrix_io/malos/cli.py
"""
malosclient -- Python MALOS client

Usage:
  malosclient [options] <driver>

Options:
  -h ADDR --malos-host=ADDR   MALOS Service host [default: localhost].
  --help                      Show help screen.
  --update-delay=VAL          Delay between data updates in seconds [default: 2.0].
  --keepalive-timeout=VAL     How long to wait for a keep alive ping before stopping data updates
                              [default: 10.0] secs.
  --version                   Show version.
  --driver-config-file=PATH   Serialized driver config protobuf file to load.
  --loglevel=LEVEL            Desired loglevel output.

MALOS Drivers (base Ports):

    IMU (port 20013) Inertial Measuring Unit.
    HUMIDITY (port 20017) Humidity sensor.
    PRESSURE (port 20025) Pressure sensor.
    UV (port 20029) Ultra Violet light sensor.
    MICARRAY_ALSA (port 20037) Microphone arrays.
    VISION (port 60001) Computer Vision engine events.
    EVERLOOP (port 20021) Led everloop (write only)

Examples:

    # Read IMU from a locally running MALOS service
    malosclient IMU

    # Read IMU from a remotely running MALOS service
    malosclient --malos-host 192.168.0.101 IMU

    # Load FACE driver config from file
    malosclient --driver-config-file ~/driver_config.proto VISION

"""
import asyncio
import logging
import math
import os
import sys

from docopt import docopt
from matrix_io.proto.malos.v1 import driver_pb2
from matrix_io.proto.malos.v1 import sense_pb2, io_pb2
from matrix_io.proto.vision.v1 import vision_pb2

from matrix_io.malos import driver
""" Driver to proto message mappings """
DRIVER_PROTOS = {
    'HUMIDITY': sense_pb2.Humidity(),
    'IMU': sense_pb2.Imu(),
    'MICARRAY_ALSA': io_pb2.MicArrayParams(),
    'PRESSURE': sense_pb2.Pressure(),
    'UV': sense_pb2.UV(),
    'VISION': vision_pb2.VisionResult()
}


async def data_handler(malos_driver, driver_name):
    """
    Sample coroutine accessing the MALOS driver data generator

    Args:
        malos_driver: MALOSDriver instance
        driver_name: driver name as a string (IMU, UV, PRESSURE)

    Returns:
        None
    """
    async for data in malos_driver.get_data():
        proto_msg = DRIVER_PROTOS[driver_name].FromString(data)

        if driver_name == 'MICARRAY_ALSA':
            print('Azimuthal angle (deg): {}'.format(
                proto_msg.azimutal_angle * 180.0 / math.pi))
            print('Polar angle (deg): {}'.format(
                proto_msg.polar_angle * 180.0 / math.pi))
        else:
            print(proto_msg)


async def status_handler(malos_driver):
    """
    Sample coroutine accessing the MALOS driver error generator

    Args:
        malos_driver: MalosDriver instance

    Returns:
        None
    """

    type_mapping = {
        driver_pb2.Status.MESSAGE_TYPE_NOT_DEFINED: "Not Defined",
        driver_pb2.Status.STARTED: "Started",
        driver_pb2.Status.STOPPED: "Stopped",
        driver_pb2.Status.CONFIG_RECEIVED: "Config Received",
        driver_pb2.Status.COMMAND_EXECUTED: "Command Executed",
        driver_pb2.Status.STATUS_CRITICAL: "Critical log",
        driver_pb2.Status.STATUS_ERROR: "Error log",
        driver_pb2.Status.STATUS_WARNING: "Warning log",
        driver_pb2.Status.STATUS_INFO: "Info log",
        driver_pb2.Status.STATUS_DEBUG: "Debug log"
    }

    async for msg in malos_driver.get_status():
        """ STDERR Error message printer """

        print(type_mapping[msg.type])

        if msg.uuid:
            print("UUID: {}".format(msg.uuid))
        if msg.message:
            print("MESSAGE: {}".format(msg.message))


def main():
    """ CLI entrypoint.

    Parses command line options and initializes the MALOS driver

    Returns:
        None
    """
    options = docopt(__doc__)

    # Driver configuration
    driver_config = driver_pb2.DriverConfig()
    driver_name = options['<driver>'].upper()

    # Sanity check on driver
    try:
        driver_port = getattr(driver, '{}_PORT'.format(driver_name))
    except AttributeError:
        print(
            "Driver '%s' is not valid, try any of: IMU, HUMIDITY, PRESSURE,"
            " UV, MICARRAY_ALSA" % options['<driver>'],
            file=sys.stderr)
        sys.exit(1)

    # Sanity checks on driver config
    try:
        update_delay = float(options['--update-delay'])
    except ValueError:
        print("Invalid --update-delay value. Try something like 1.3")
        sys.exit(1)
    else:
        # Set the delay between updates that the driver returns
        driver_config.delay_between_updates = update_delay

    try:
        keepalive_timeout = float(options['--keepalive-timeout'])
    except ValueError:
        print("Invalid --keepalive-delay value. Try something like 8.0")
        sys.exit(1)
    else:
        # Stop sending updates if there is no ping for 10 seconds
        driver_config.timeout_after_last_ping = keepalive_timeout

    if options['--driver-config-file'] is not None:
        try:
            file_content = open(
                os.path.expanduser(options['--driver-config-file']),
                'rb').read()
        # @TODO (heitorgo, maciekrb) let's improve for specific exceptions here
        except Exception as err:
            print("Failed to load driver config file.", file=sys.stderr)
            sys.exit(1)
        else:
            driver_config.ParseFromString(file_content)

    # Logging configuration
    if options['--loglevel']:
        numeric_level = getattr(logging, options['--loglevel'].upper(), None)
        if not isinstance(numeric_level, int):
            logging.error('Invalid --loglevel: %s', options['--loglevel'])
            raise SystemExit(1)

        logging.basicConfig(format='%(asctime)s %(message)s', level=numeric_level)


    malos_driver = driver.MalosDriver(options['--malos-host'], driver_port)

    loop = asyncio.get_event_loop()
    
    # Schedule tasks
    loop.run_until_complete(malos_driver.configure(driver_config))
    loop.create_task(malos_driver.start_keep_alive())
    loop.create_task(data_handler(malos_driver, driver_name))
    loop.create_task(status_handler(malos_driver))
    try:
        loop.run_forever()
    except KeyboardInterrupt:
        print('Shutting down. Bye, bye !', file=sys.stderr)
    finally:
        loop.stop()
        asyncio.gather(*asyncio.Task.all_tasks()).cancel()

        loop.run_until_complete(loop.shutdown_asyncgens())
        loop.close()
